File: Models/BlogPost.cs
namespace laoyu_blog_backend.Models;

public class BlogPost
{
    public int Id { get; set; }

    public string Slug { get; set; } = string.Empty;

    public string Title { get; set; } = string.Empty;

    public ICollection<Category> Categories { get; set; }
     = new List<Category>();

    public string Content { get; set; } = string.Empty;

    public BlogPostStatus Status { get; set; }
        = BlogPostStatus.Draft;

    public DateTime? PublishedAtUtc { get; set; }

    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;

    public bool Publish()
    {
        if (Status == BlogPostStatus.Published
            && PublishedAtUtc is not null)
        {
            return false;
        }

        Status = BlogPostStatus.Published;
        PublishedAtUtc = DateTime.UtcNow;

        return true;
    }

    public bool Unpublish()
    {
        if (Status == BlogPostStatus.Draft
            && PublishedAtUtc is null)
        {
            return false;
        }

        Status = BlogPostStatus.Draft;
        PublishedAtUtc = null;

        return true;
    }
}

File: frontend/src/types/blogPost.ts
export interface Category {
  id: number
  name: string
  slug: string
}

export type BlogPostStatus = 'Draft' | 'Published'

export interface BlogPost {
  id: number
  slug: string
  title: string
  content: string
  categories: Category[]
  tags: string[]
  status: BlogPostStatus
  publishedAtUtc: string | null
  createdAtUtc: string
}

export interface BlogPostInput {
  title: string
  slug: string
  content: string
  categoryNames: string[]
  tags: string[]
}

export interface PagedResult<T> {
  items: T[]
  page: number
  pageSize: number
  totalPages: number
  totalItems: number
}
